packages/astro/src/events/error.ts

```diff
@@ -65,11 +66,47 @@ export function eventError({
 	isFatal: boolean;
 }): { eventName: string; payload: ErrorEventPayload }[] {
 	const payload: ErrorEventPayload = {
-		code: err.code || AstroErrorData.UnknownError.code,
+		code: err.code || err.errorCode || AstroErrorData.UnknownError.code,
+		name: err.name,
 		plugin: err.plugin,
 		cliCommand: cmd,
 		isFatal: isFatal,
-		anonymousMessageHint: anonymizeErrorMessage(err.message),
+		anonymousMessageHint:
+			AstroError.is(err) && err.errorCode
+				? getSafeErrorMessage((getErrorDataByCode(err.errorCode)?.data as any).message)
+				: anonymizeErrorMessage(err.message),
 	};
 	return [{ eventName: EVENT_ERROR, payload }];
 }
+
+/**
+ * Safely get the error message from an error, even if it's a function.
+ */
+// eslint-disable-next-line @typescript-eslint/ban-types
+function getSafeErrorMessage(message: string | Function): string {
```

Comment:
The logic here is the same one that generates the [error reference](https://docs.astro.build/en/reference/error-reference/) on the docs, which has been proven to work:tm: